pywbem_mock/_providerdispatcher.py

```diff
@@ -95,207 +97,580 @@ def default_method_provider(self):
+                        "not exist in namespace {1!A} of the CIM repository.",
+                        InstanceName.classname, namespace))
 
-        instance_store = self.cimrepository.get_instance_store(namespace)
+        # Verify that the instance to be deleted exists in the CIM repository.
         if not instance_store.object_exists(InstanceName):
             raise CIMError(
                 CIM_ERR_NOT_FOUND,
-                _format("Instance {0!A} not found in CIM repository",
+                _format("Instance to be deleted does not exist in the CIM "
+                        "repository: {0!A}",
                         InstanceName))
 
+        # Determine the provider to be used. Note that a registered provider
+        # always has all responder methods for the provider type, either
+        # implemented or inherited.
         provider = self.provider_registry.get_registered_provider(
-            InstanceName.namespace, 'instance-write', InstanceName.classname)
+            namespace, 'instance-write', InstanceName.classname)
+        if not provider:
+            provider = self.default_instance_write_provider
 
-        # Execute the method in the provider if the method exists. Otherwise
-        # fall back to the default provider
-        if provider:
-            try:
-                return provider.DeleteInstance(InstanceName)
-            except AttributeError:
-                pass
+        # Call the provider method.
+        result = provider.DeleteInstance(InstanceName)
 
-        return self.default_instance_write_provider.DeleteInstance(
-            InstanceName)
+        # Verify provider method result.
+        assert result is None
 
-    def InvokeMethod(self, namespace, methodname, objectname, Params):
+    def InvokeMethod(self, methodname, localobject, params):
         # pylint: disable=invalid-name
         """
-        Default Method provider.
-        NOTE: There is no default method provider because all method
-        providers provide specific functionality and there is no way
-        to do that in a default method provider.
+        Dispatcher for the InvokeMethod provider method.
+
+        This method performs validations and if successful, routes the provider
+        method call either to a registered provider, or to the default provider.
+
+        Parameters:
+
+          methodname (string): Method name
+
+          localobject (CIMInstanceName or CIMClassName): Target object, with
+            namespace set. Types are validated.
+
+          params (NocaseDict): Input parameters, as follows:
+            * key (string): Parameter name.
+            * value (CIMParameter): Parameter value.
+            Types are validated.
+
+        Returns:
+
+          A tuple of (returnvalue, outparams), with these tuple items:
+
+            * returnvalue (CIM data type): Return value.
+
+            * outparams (NocaseDict): Output parameters, with:
```

Comment:
How does a user in an environment like pywbemcli use NocaseDict since that is a pywbem internal?  I understand why we have dict on input parameters, name case insensitivity but wouldn't just list of CIMParameter be easier???

Reply:
NOISSUE
Note, this is the return value of the dispatcher method. In the user-defined provider, the user does not need to create a NocaseDict. In case the output parameters are returned as dict, that is a standard dict.

Using a NocaseDict is different, that happens all the time and we describe that in the pywbem docs. But the user never needs to import the NocaseDict name.